kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/InBandKeyManager.java

```diff
@@ -154,27 +144,52 @@ private Supplier<CompletionStage<KeyContext>> makeKeyContext(@NonNull K kekId) {
     public CompletionStage<Void> encrypt(@NonNull EncryptionScheme<K> encryptionScheme,
                                          @NonNull List<? extends Record> records,
                                          @NonNull Receiver receiver) {
-
-        var fieldsHeader = createEncryptedFieldsHeader(encryptionScheme.recordFields());
-        return currentDekContext(encryptionScheme.kekId(), records.size()).thenAccept(keyContext -> {
-            var dekHeader = createEdekHeader(keyContext);
-            var maxValuePlaintextSize = encryptionScheme.recordFields().contains(RecordField.RECORD_VALUE)
-                    ? records.stream().mapToInt(Record::valueSize).max().orElse(-1)
-                    : -1;
+        return currentDekContext(encryptionScheme.kekId()).thenCompose(keyContext -> {
+            // access to the key context is synchronized
             synchronized (keyContext) {
-                var valueCiphertext = maxValuePlaintextSize >= 0 ? bufferPool.acquire(keyContext.encodedSize(maxValuePlaintextSize)) : null;
-                try {
-                    encryptRecords(encryptionScheme, keyContext, records, fieldsHeader, dekHeader, valueCiphertext, receiver);
-                }
-                finally {
-                    if (maxValuePlaintextSize >= 0) {
-                        bufferPool.release(valueCiphertext);
+                // if it's not alive we know a previous encrypt call has replaced the stage in the cache and fall through to retry encrypt
+                if (keyContext.isAlive()) {
+                    if (!keyContext.hasAtLeastRemainingEncryptions(records.size())) {
+                        // replace the key context stage in the cache, then call encrypt again
+                        rotateKeyContext(encryptionScheme, keyContext);
+                    }
+                    else {
+                        return encryptUsingSharedCiphertextBuffer(encryptionScheme, records, receiver, keyContext);
                     }
                 }
             }
+            // todo add a recursion limit
+            return encrypt(encryptionScheme, records, receiver);
         });
     }
 
+    @NonNull
+    private CompletableFuture<Void> encryptUsingSharedCiphertextBuffer(@NonNull EncryptionScheme<K> encryptionScheme, @NonNull List<? extends Record> records,
```

Comment:
I think the fact this method uses a shared temporary buffer is an implementation concern.  I don't think having it in the name helps understandability.

Reply:
Yep I don't think it helps here precisely, but I was casting around trying to make it clearer that the buffers the `Receiver` receives must be consumed during the `Received#accept()` call as they can be a shared resource. It feels like the sharedness is part of the Receiver API currently.

Reply:
Sounds more like it should go in the `Receiver` javadoc that the buffer may be shared and the user should not expect it's contents to be stable.

Reply:
also thought it should be documented somewhere like KeyManager something about the ordering guarantees, like `Receiver` will receive the encrypted records in the original order and can expect no parallel calls.